### ISSUE TITLE
[Question] Integration in mainstream

### ISSUE THREAD

**Issue by carsanbu** [bebenlebricolo/CMake-AtmelStudio7-compatibility#43]
There are plans to integrate this fork in mainstream? Would be awesome to have this feature in the Kitware cmake and more people could contribute to the project.

**Comment by bebenlebricolo** [comment]
Hi @carsanbu, I've raised this question some time ago in the Cmake discourse forums, however as it seemed to me (at the time) that Microchip was easing AtmelStudio out of support, the guys from the CMake main development told me they did not like the idea to integrate support for a dying IDE/platform. 
I completely understand their approach, even if I don't really like it. That's why I decided to fork CMake's code and provide regular mainstream CMake updates to it ; this is far from perfect though..

You can find this [discussion here](https://discourse.cmake.org/t/generator-writing-a-new-generator-for-atmelstudio-7-ide/1818).

In the long run, it appears that Microchip is somehow still maintaining AtmelStudio under a new name, so the project is still quite alive. However, assuming I was alone on this project I decided to substantially diverge the implementation from what CMake does in the source code (with some weird choices I made that made sense to me at that time).

We could reactivate the discussion though, telling that it appears Microchip is finally keeping AtmelStudio alive ; however I'm not sure this implementation will ever be ported to mainstream CMake.

**Comment by carsanbu** [comment]
Thank you so much for your replies, I have write to the discussions with the aim to reactivate it.

**Comment by bebenlebricolo** [comment]
Hi @carsanbu , thank you for your support, it means a lot to me as I did not think this fork will get much attention back when I worked on it.

As you might have noticed, we had a small chat with the guys of CMake and it seems they are not quite happy with the approach I took with external dependencies (PugiXML and GoogleTests).
I can see several roadblocks at the moments :
* External dependencies are not quite compliant with the Cmake's philosophy ; instead its development community tend to absorb the source code directly and rename it under the name "utilities/cm***" which I personally don't really like.
* The developers help/dev/source.rst file, standing as a development guide, says in essence that the code should compile with C++11, using some cmake-custom-made wrappers (Utilities/cmext/algorithm for instance) which further reduces the ease of development with modern C++. This is an issue (to me) as the code I wrote uses recent C++ features, and I know that they have a bot on gitlab that rejects all PRs until they match the requirements
* I broke the formatting pass intentionally (the git hooks were enforcing a 80 characters max length formatting ; I don't quite like to code on a Gameboy screen so I removed it :smile: )
* GoogleTest reverting to CTests appears to me as some kind of regression. For what I saw in the cmake codebase itself, CTest does not come with the same features as GoogleTests, and even it's is perfectly doable to revert to CTest based-operation, I don't quite like the loss in development ease and overal quality that implies. That's of course a personal opinion after all.
* Github cannot do PR's over the void of internet into Gitlab (and that's sad).
* I modified the directory layout of AtmelStudio compatibility layer quite a lot compared to every other components of CMake (which are all dumped in the same Source/** folder, in a flat-tree manner). This is a minor change.

I think you've got it, CMake has its own development philosophy. It's an "old" software, with tons of rules and I totally get it : it's necessary for a long-lived project to provide a frame, and solid rules to prevent people (like me) to create breaches and compromise the software's consistency. It could be a real mess in the end...
However I'm not really happy as to how it turns out. 

I feel backporting the AtmelStudio compatibility layer to mainstream CMake requires a lot of work.
And I'm not sure it is essential right now ; I think we can live with a custom version of CMake ; and in facts this compatibility layer is not really stable neither.

I think I'd better continue improving this version, porting new CMake stable releases to this fork as time passes, fixing the bugs  and adding more features and meaning to this project. 
As a bottom line, I'll be really happy to help anyone who wants to delve into the implementation details, for future developments.

I also think that we might consider decoupling from the standard CMake implementation even more (I really don't like the fact that we are relying on *.cmake scripts to find targets within an OS's file-system. I get it's a flexible approach to deal with low-level details, but it does not supersedes good old C++ code and is really awful when debugging).
In the end, this version could either strip-out a lot of Cmake's internals, or keep them to provide a "CMake + AS7 compatibility layer", which is the choice I made until now.

**Comment by carsanbu** [comment]
Even if it is not integrated in mainstream this is a great project that can be CMake compatible as clang is compatible to gcc only in interface.
Having regard to the problems with integrating new functionality to CMake maybe my next projects I will give a try to meson build.
Thank you so much!

P.S. 80 character is the best to view multiple files at same time 😄

**Comment by bebenlebricolo** [comment]
I'm afraid I did not get the part with clang/gcc, could you develop on it ? To my knowledge CMake is not tied to any specific compiler and builds with whatever you have at hands (msvc, gcc, clang, and probably other ones). This extends to this project as well, of course.
However I think you are speaking about something else though, so feel free to comment on this 😄.
More than a CMake compatible project, this project **really is** CMake + dedicated generators and tools after all.

Regarding the new functionalities, this project suffers from lack of time (as most open source project developed using spare time) and amongst other things misses proper end-to-end tests.
I have several unit-tests that assert the behavior of compiler flag parsing for instance, but none that can carry the whole workflow.

And of course I warmly welcome anyone that has the will to help with some features, debug or anything else ; documentation as well.

**Comment by carsanbu** [comment]
Maybe clang and gcc are not the best example, but I mean although clang use new tools like LLVM to build they continue to using the same input interface in it's CLI, like -Werror or -Wno-foo, then you can esasily change from one to another.

**Comment by bebenlebricolo** [comment]
> [...] but I mean although clang use new tools like LLVM to build they continue to using the same input interface in it's CLI, like -Werror or -Wno-foo, then you can esasily change from one to another.

I generally agree with the idea, however I don't quite get the message ; are you speaking about the way CMake itself works ? Or the AtmelStudio7Generator works ? 

When I talk about flag parsing, that's the AtmelStudio7Generator which is involved.
One of its task is to retrieve the `CMAKE_CXX_FLAGS`and `CMAKE_C_FLAGS` from the build tree decription in `CMakeLists.txt`,  parse the flags and categorize them.
This can be seen in this test file : [testCmAvrGccCompiler.cpp](https://github.com/bebenlebricolo/CMake-AtmelStudio7-compatibility/blob/master/Tests/CMakeLib/testCmAtmelStudioTools/testCmAvrGccCompiler.cpp).
It basically verifies that the compiler abstraction I wrote for avr-gcc is compliant with the flags supported by AtmelStudio7.
Here is one the abstractions I'm speaking about : [AvrGCC8Toolchain.h](https://github.com/bebenlebricolo/CMake-AtmelStudio7-compatibility/blob/master/Source/AtmelStudio7Generators/AS7Toolchains/AvrGCC8Toolchain.h).

So as the only toolchain for which I provided support is avr gcc 8 bit only (even worse : it's still lacking c++ support).
As such, I have several components whose role is to correctly identify the flags that are natively supported by AtmelStudio (with their ui pages, in project properties) and the flags that should appear as "other flags".

So if it happens that we have a clang-based toolchain for devices supported by AtmelStudio, then eventually we'll need to duplicate part of the flag support already supported by the GCC abstraction I wrote for that purpose.
